core/modules/replica/Job.h

```diff
@@ -0,0 +1,482 @@
+
+        /// The job has been constructed, and no attempt to execute it has
+        /// been made.
+        CREATED,
+
+        /// The job is in a progress
+        IN_PROGRESS,
+
+        /// The job is finihed. See extended status for more details
+        /// (the completion status, etc.)
+        FINISHED
+    };
+
+    /// Return the string representation of the primary state
+    static std::string state2string(State state) ;
+
+    /// Refined public sub-state of the job once it's FINISHED as per
+    /// the above defined primary state.
+    enum ExtendedState {
+
+        /// No extended state exists at this time
+        NONE,
+
+        /// The job has been fully implemented
+        SUCCESS,
+
+        /// Problems with job configuration found
+        CONFIG_ERROR,
+
+        /// The job has failed
+        FAILED,
+
+        /// Qserv notification failed
+        QSERV_FAILED,
+
+        /// Qserv reported that the source chunk is in use and couldn't be removed
+        QSERV_IN_USE,
+
+        /// Expired due to a timeout (as per the Configuration)
+        EXPIRED,
+
+        /// Explicitly cancelled on the client-side (similar to EXPIRED)
+        CANCELLED
+    };
+
+    /// @return string representation of the extended state
+    static std::string state2string(ExtendedState state) ;
+
+    /// @return string representation of the combined state
+    static std::string state2string(State state, ExtendedState extendedState) {
+        return state2string(state) + "::" +state2string(extendedState);
+    }
+
+    /// The job options container
+    struct Options {
+
+        /// the priority level
+        int  priority;
+
+        /// the flag indicating of this job can't be run simultaneously
+        /// along with other jobs.
+        bool exclusive;
+
+        /// the flag indicating if the job is allowed to be interrupted by some
+        /// by other jobs.
+        bool preemptable;
+    };
+
+    // Default construction and copy semantics are prohibited
+
+    Job() = delete;
+    Job(Job const&) = delete;
+    Job& operator=(Job const&) = delete;
+
+    virtual ~Job() = default;
+
+    /// @return a reference to the Controller,
+    Controller::Ptr controller() const { return _controller; }
+
+    /// @return the optional identifier of a parent job
+    std::string const& parentJobId() const { return _parentJobId; }
+
+    /// @return a string representing a type of a job.
+    std::string const& type() const { return _type; }
+
+    /// @return a unique identifier of the job
+    std::string const& id() const { return _id; }
+
+    /// @return the primary status of the job
+    State state() const { return _state; }
+
+    /// @return the extended state of the job when it's finished
+    ExtendedState extendedState() const { return _extendedState; }
+
+    /// @return string representation of the combined state of the object
+    std::string state2string() const { return state2string(_state, _extendedState); }
```

Comment:
based on my understanding from previous discussions, I think this may not be thread safe. Probably not catastrophically so, but it may be confusing in some cases.

Reply:
Reinforced to be thread-safe